README.md
# README <!-- omit in toc -->

`quick-roueter` is a REST API utility that helps configure routes by moving all the boilerplate into a simple interface

[changelog](CHANGELOG.md)

- [Why use quick-router](#why-use-quick-router)
- [Supported Server Packages](#supported-server-packages)
- [Installation](#installation)

## Why use quick-router

building REST APIs can come with a lot of bolierplate

you might end up re-implementing the same helpers over and over again or worse, have lazy copy-pasta sprinkled throughout your codebase

the goal of `quick-router` is to remove as much bolierplate as possible, while still providing avenues for customization

## Supported Server Packages

at this time, `express` is the only supported server package

want more? open an issue or submit a pull request

## Installation

``` bash
npm install express @dallen.baldwin/quick-router
```


CHANGELOG.md
# CHANGELOG <!-- omit in toc -->

This project does its best to adhere to [semantic versioning](https://semver.org/) and follows a similar format to [keepachangelog](https://keepachangelog.com/en/1.0.0/)

- [0.0.1](#001)

## [0.0.1][001]

[001]: https://github.com/dallenbaldwin/quick-router "0.0.1"


src/tests/index.ts
import express, { json } from 'express';
import { router } from './router';

const testingApp = () => {
  const app = express();
  express();
  app.use(json());
  app.use('/api', router);
  const port = 3000;
  app.listen(port, () => {
    console.log(`Test app listening at http://localhost:${port}/api`);
  });
};

export { testingApp };


src/test/data/index.ts
import { faker } from '@faker-js/faker';
import { Customer } from './models/Customer';
import { Order } from './models/Order';
import { OrderItem } from './models/OrderItem';
import { Part } from './models/Part';
import { Product } from './models/Product';

interface Database {
  customers: Customer[];
  orders: Order[];
  orderItems: OrderItem[];
  parts: Part[];
  products: Product[];
}

const database: Database = {
  customers: [],
  orders: [],
  orderItems: [],
  parts: [],
  products: [],
};

const entities = <T>(mapper: () => T, count: number): T[] =>
  Array(count)
    .fill(undefined)
    .map(() => mapper());

database.customers = entities(() => {
  const first = faker.name.firstName();
  const last = faker.name.lastName();
  return new Customer({
    first,
    last,
    city: faker.address.city(),
    state: faker.address.state(),
    email: faker.internet.email(first, last),
    street: faker.address.streetAddress(),
    street2: faker.address.secondaryAddress(),
    zip: faker.address.zipCode(),
    phone: faker.phone.number('(###) ###-###'),
  });
}, 20);

database.parts = entities(
  () =>
    new Part({
      cost: +faker.commerce.price(),
      description: faker.commerce.productDescription(),
      number: `${faker.datatype.number()}-${faker.commerce.productMaterial()}`,
    }),
  20,
);

database.products = entities(
  () =>
    new Product({
      part: faker.helpers.arrayElement(database.parts),
      price: +faker.commerce.price(),
      description: faker.commerce.productDescription(),
      number: `${faker.datatype.number()}-${faker.commerce.productName()}`,
    }),
  20,
);

database.orders = entities(
  () =>
    new Order({
      customer: faker.helpers.arrayElement(database.customers),
      dateToFulfill: faker.datatype.datetime(),
      number: `${faker.random.alphaNumeric()}-${faker.datatype.number()}`,
    }),
  10,
);

database.orderItems = entities(
  () =>
    new OrderItem({
      order: faker.helpers.arrayElement(database.orders),
      price: +faker.commerce.price(),
      product: faker.helpers.arrayElement(database.products),
      qtyToFulfill: faker.datatype.number(),
    }),
  50,
);

// inverse relations
database.orders = database.orders.map((order) => {
  const { id } = order;
  order.orderItems = database.orderItems.filter(
    ({ order: { id: orderId } }) => id === orderId,
  );
  return order;
});

database.customers = database.customers.map((customer) => {
  const { id } = customer;
  customer.orders = database.orders.filter(
    ({ customer: { id: customerId } }) => id === customerId,
  );
  return customer;
});

database.parts = database.parts.map((part) => {
  const { id } = part;
  part.products = database.products.filter(
    ({ part: { id: partId } }) => id === partId,
  );
  return part;
});

database.products = database.products.map((product) => {
  const { id } = product;
  product.orderItems = database.orderItems.filter(
    ({ product: { id: productId } }) => id === productId,
  );
  return product;
});

export { database };


src/test/router/index.ts
import { Router } from 'express';

const router = Router();
export { router };


src/constants/errors.ts
/**
 * Copyright Dallen Baldwin 2022 - now
 * Distributed under the Boost Software License, Version 1.0.
 * (See accompanying file ./LICENSE or copy at https://www.boost.org/LICENSE_1_0.txt)
 */

/**
 * A union type that encompasses all error codes
 */
export type ErrorCode = ClientError | ServerError;

/**
 * Names, codes, and descriptions take from [mdn](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#server_error_responses)
 */
export enum ClientError {
  /**
   * The server cannot or will not process the request due to
   * something that is perceived to be a client error
   *
   * (e.g., malformed request syntax, invalid request message
   * framing, or deceptive request routing).
   */
  BAD_REQUEST = '400',
  /**
   * Although the HTTP standard specifies "unauthorized",
   * semantically this response means "unauthenticated".
   *
   * That is, the client must authenticate itself to get the
   * requested response.
   */
  UNAUTHORIZED = '401',
  /**
   * _Experimental_: expect behavior to change in the future.
   *
   * This response code is reserved for future use.
   *
   * The initial aim for creating this code was using it for
   * digital payment systems, however this status code is used
   * very rarely and no standard convention exists.
   */
  PAYMENT_REQUIRED = '402',
  /**
   * The client does not have access rights to the content; that is,
   * it is unauthorized, so the server is refusing to give the
   * requested resource.
   *
   * Unlike 401 Unauthorized, the client's identity is known to
   * the server.
   */
  FORBIDDEN = '403',
  /**
   * The server can not find the requested resource.
   *
   * In the browser, this means the URL is not recognized.
   *
   * In an API, this can also mean that the endpoint is valid
   * but the resource itself does not exist.
   *
   * Servers may also send this response instead of 403 Forbidden
   * to hide the existence of a resource from an unauthorized client.
   *
   * This response code is probably the most well known due to
   * its frequent occurrence on the web.
   */
  NOT_FOUND = '404',
  /**
   * The request method is known by the server but is not
   * supported by the target resource.
   *
   * For example, an API may not allow calling DELETE to remove
   * a resource.
   */
  METHOD_NOT_ALLOWED = '405',
  /**
   * This response is sent when the web server, after
   * performing server-driven content negotiation, doesn't find
   * any content that conforms to the criteria given by the user agent.
   */
  NOT_ACCEPTABLE = '406',
  /**
   * This is similar to 401 Unauthorized but authentication is
   * needed to be done by a proxy.
   */
  PROXY_AUTHENTICATION_REQUIRED = '407',
  /**
   * This response is sent on an idle connection by some servers,
   * even without any previous request by the client.
   *
   * It means that the server would like to shut down this
   * unused connection.
   *
   * This response is used much more since some browsers,
   * like Chrome, Firefox 27+, or IE9, use HTTP pre-connection
   * mechanisms to speed up surfing.
   *
   * Also note that some servers merely shut down the
   * connection without sending this message.
   */
  REQUEST_TIMEOUT = '408',
  /**
   * This response is sent when a request conflicts with the
   * current state of the server.
   */
  CONFLICT = '409',
  /**
   * This response is sent when the requested content
   * has been permanently deleted from server, with no
   * forwarding address.
   *
   * Clients are expected to remove their caches and links
   * to the resource.
   *
   * The HTTP specification intends this status code to be used
   * for "limited-time, promotional services".
   *
   * APIs should not feel compelled to indicate resources
   * that have been deleted with this status code.
   */
  GONE = '410',
  /**
   * Server rejected the request because the Content-Length
   * header field is not defined and the server requires it.
   */
  LENGTH_REQUIRED = '411',
  /**
   * The client has indicated preconditions in its headers
   * which the server does not meet.
   */
  PRECONDITION_FAILED = '412',
  /**
   * Request entity is larger than limits defined by server.
   *
   * The server might close the connection or return an Retry-After
   * header field.
   */
  PAYLOAD_TOO_LARGE = '413',
  /**
   * The URI requested by the client is longer than the server
   * is willing to interpret.
   */
  URI_TOO_LONG = '414',
  /**
   * The media format of the requested data is not supported by
   * the server, so the server is rejecting the request.
   */
  UNSUPPORTED_MEDIA_TYPE = '415',
  /**
   * The range specified by the Range header field in the request
   * cannot be fulfilled.
   *
   * It's possible that the range is outside the size of the target
   * URI's data.
   */
  RANGE_NOT_SATISFIABLE = '416',
  /**
   * This response code means the expectation indicated by the
   * Expect request header field cannot be met by the server.
   */
  EXPECTATION_FAILED = '417',
  /**
   * The server refuses the attempt to brew coffee with a teapot.
   */
  I_AM_A_TEAPOT = '418',
  /**
   * The request was directed at a server that is not able to
   * produce a response.
   *
   * This can be sent by a server that is not configured to
   * produce responses for the combination of scheme and authority
   * that are included in the request URI.
   */
  MISDIRECTED_REQUEST = '421',
  /**
   * [WebDAV](https://developer.mozilla.org/en-US/docs/Glossary/WebDAV)
   *
   * The request was well-formed but was unable to be followed
   * due to semantic errors.
   */
  UNPROCESSABLE_ENTITY = '422',
  /**
   * [WebDAV](https://developer.mozilla.org/en-US/docs/Glossary/WebDAV)
   *
   * The resource that is being accessed is locked.
   */
  LOCKED = '423',
  /**
   * [WebDAV](https://developer.mozilla.org/en-US/docs/Glossary/WebDAV)
   *
   * The request failed due to failure of a previous request.
   */
  FAILED_DEPENDENCY = '424',
  /**
   * _Experimental_: expect behavior to change in the future.
   *
   * Indicates that the server is unwilling to risk processing a
   * request that might be replayed.
   */
  TOO_EARLY = '425',
  /**
   * The server refuses to perform the request using the current
   * protocol but might be willing to do so after the client
   * upgrades to a different protocol.
   *
   * The server sends an Upgrade header in a 426 response to
   * indicate the required protocol(s).
   */
  UPGRADE_REQUIRED = '426',
  /**
   * The origin server requires the request to be conditional.
   *
   * This response is intended to prevent the 'lost update' problem,
   * where a client GETs a resource's state, modifies it and PUTs
   * it back to the server, when meanwhile a third party has
   * modified the state on the server, leading to a conflict.
   */
  PRECONDITION_REQUIRED = '428',
  /**
   * The user has sent too many requests in a given amount of
   * time ("rate limiting").
   */
  TOO_MANY_REQUESTS = '429',
  /**
   * The server is unwilling to process the request because
   * its header fields are too large.
   *
   * The request may be resubmitted after reducing the size of
   * the request header fields.
   */
  REQUEST_HEADER_FIELDS_TOO_LARGE = '431',
  /**
   * The user agent requested a resource that cannot legally be
   * provided, such as a web page censored by a government.
   */
  UNAVAILABLE_FOR_LEGAL_REASONS = '451',
}

/**
 * names, codes, and descriptions take from [mdn](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#server_error_responses)
 */
export enum ServerError {
  /**
   * The server has encountered a situation it does not know
   * how to handle.
   */
  INTERNAL_SERVER_ERROR = '500',
  /**
   * The request method is not supported by the server
   * and cannot be handled.
   *
   * The only methods that servers are required to support
   * (and therefore that must not return this code) are GET and HEAD.
   */
  NOT_IMPLEMENTED = '501',
  /**
   * This error response means that the server, while
   * working as a gateway to get a response needed to handle
   * the request, got an invalid response.
   */
  BAD_GATEWAY = '502',
  /**
   * The server is not ready to handle the request.
   *
   * Common causes are a server that is down for maintenance
   * or that is overloaded.
   *
   * Note that together with this response, a user-friendly
   * page explaining the problem should be sent.
   *
   * This response should be used for temporary conditions
   * and the Retry-After HTTP header should, if possible,
   * contain the estimated time before the recovery of the service.
   *
   * The webmaster must also take care about the caching-related
   * headers that are sent along with this response, as these
   * temporary condition responses should usually not be cached.
   */
  SERVICE_UNAVAILABLE = '503',
  /**
   * This error response is given when the server is acting
   * as a gateway and cannot get a response in time.
   */
  GATEWAY_TIMEOUT = '504',
  /**
   * The HTTP version used in the request is not supported by
   * the server.
   */
  HTTP_VERSION_NOT_SUPPORTED = '505',
  /**
   * The server has an internal configuration error: the chosen
   * variant resource is configured to engage in transparent
   * content negotiation itself, and is therefore not a proper
   * end point in the negotiation process.
   */
  VARIANT_ALSO_NEGOTIATES = '506',
  /**
   * [WebDAV](https://developer.mozilla.org/en-US/docs/Glossary/WebDAV)
   *
   * The method could not be performed on the resource
   * because the server is unable to store the representation
   * needed to successfully complete the request.
   */
  INSUFFICIENT_STORAGE = '507',
  /**
   * [WebDAV](https://developer.mozilla.org/en-US/docs/Glossary/WebDAV)
   *
   * The server detected an infinite loop while processing the request.
   */
  LOOP_DETECTED = '508',
  /**
   * Further extensions to the request are required for the
   * server to fulfill it.
   */
  NOT_EXTENDED = '510',
  /**
   * Indicates that the client needs to authenticate to gain
   * network access.
   */
  NETWORK_AUTHENTICATION_REQUIRED = '511',
}


src/constants/successes.ts
/**
 * Copyright Dallen Baldwin 2022 - now
 * Distributed under the Boost Software License, Version 1.0.
 * (See accompanying file ./LICENSE or copy at https://www.boost.org/LICENSE_1_0.txt)
 */

/**
 * A union type that encompasses all successful response codes
 */
export type SuccessCode = Information | Success | Redirection;

/**
 * Names, codes, and descriptions take from [mdn](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#server_error_responses)
 */
export enum Information {
  /**
   * This interim response indicates that the client should continue
   * the request or ignore the response if the request is already
   * finished.
   */
  CONTINUE = '100',
  /**
   * This code is sent in response to an Upgrade request
   * header from the client and indicates the protocol the server
   * is switching to.
   */
  SWITCHING_PROTOCOLS = '101',
  /**
   * [WebDAV](https://developer.mozilla.org/en-US/docs/Glossary/WebDAV)
   *
   * This code indicates that the server has received and is
   * processing the request, but no response is available yet.
   */
  PROCESSING = '102',
  /**
   * This status code is primarily intended to be used with the
   * Link header, letting the user agent start preloading resources
   * while the server prepares a response.
   */
  EARLY_HINTS = '103',
}

/**
 * names, codes, and descriptions take from [mdn](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#server_error_responses)
 */
export enum Success {
  /**
   * The request succeeded. The result meaning of "success" depends
   * on the HTTP method:
   *
   * - `GET`: The resource has been fetched and transmitted in the message body.
   * - `HEAD`: The representation headers are included in the response without any message body.
   * - `PUT` or `POST`: The resource describing the result of the action is transmitted in the message body.
   * - `TRACE`: The message body contains the request message as received by the server.
   */
  OK = '200',
  /**
   * The request succeeded, and a new resource was created as a result.
   *
   * This is typically the response sent after POST requests,
   * or some PUT requests.
   */
  CREATED = '201',
  /**
   * The request has been received but not yet acted upon.
   *
   * It is noncommittal, since there is no way in HTTP to later send
   * an asynchronous response indicating the outcome of the request.
   *
   * It is intended for cases where another process or server handles
   * the request, or for batch processing.
   */
  ACCEPTED = '202',
  /**
   * This response code means the returned metadata is not exactly
   * the same as is available from the origin server, but is collected
   * from a local or a third-party copy.
   *
   * This is mostly used for mirrors or backups of another resource.
   *
   * Except for that specific case, the 200 OK response is preferred
   * to this status.
   */
  NON_AUTHORITATIVE_INFORMATION = '203',
  /**
   * There is no content to send for this request, but the headers
   * may be useful.
   *
   * The user agent may update its cached headers for this resource
   * with the new ones.
   */
  NO_CONTENT = '204',
  /**
   * Tells the user agent to reset the document which sent this request.
   */
  RESET_CONTENT = '205',
  /**
   * This response code is used when the Range header is sent from
   * the client to request only part of a resource.
   */
  PARTIAL_CONTENT = '206',
  /**
   * [WebDAV](https://developer.mozilla.org/en-US/docs/Glossary/WebDAV)
   *
   * Conveys information about multiple resources, for situations
   * where multiple status codes might be appropriate.
   */
  MULTI_STATUS = '207',
  /**
   * [WebDAV](https://developer.mozilla.org/en-US/docs/Glossary/WebDAV)
   *
   * Used inside a <dav:propstat> response element to avoid repeatedly
   * enumerating the internal members of multiple bindings to the
   * same collection.
   */
  ALREADY_REPORTED = '208',
  /**
   * [HTTP Delta encoding](https://datatracker.ietf.org/doc/html/rfc3229)
   *
   * The server has fulfilled a GET request for the resource, and
   * the response is a representation of the result of one or more
   * instance-manipulations applied to the current instance.
   */
  IM_USED = '226',
}

/**
 * names, codes, and descriptions take from [mdn](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#server_error_responses)
 */
export enum Redirection {
  /**
   * The request has more than one possible response.
   *
   * The user agent or user should choose one of them.
   *
   * (There is no standardized way of choosing one of the responses,
   * but HTML links to the possibilities are recommended so the user
   * can pick.)
   */
  MULTIPLE_CHOICES = '300',
  /**
   * The URL of the requested resource has been changed permanently.
   *
   * The new URL is given in the response.
   */
  MOVED_PERMANENTLY = '301',
  /**
   * This response code means that the URI of requested resource
   * has been changed temporarily.
   *
   * Further changes in the URI might be made in the future.
   *
   * Therefore, this same URI should be used by the client in
   * future requests.
   */
  FOUND = '302',
  /**
   * The server sent this response to direct the client to get the
   * requested resource at another URI with a `GET` request.
   */
  SEE_OTHER = '303',
  /**
   * This is used for caching purposes.
   *
   * It tells the client that the response has not been modified,
   *
   * so the client can continue to use the same cached version of the
   * response.
   */
  NOT_MODIFIED = '304',
  /**
   * Defined in a previous version of the HTTP specification to
   * indicate that a requested response must be accessed by a proxy.
   *
   * @deprecated
   * deprecated due to security concerns regarding in-band configuration
   * of a proxy.
   */
  USE_PROXY = '305',
  /**
   * @deprecated
   * This response code is no longer used; it is just reserved.
   *
   * It was used in a previous version of the HTTP/1.1 specification.
   */
  UNUSED = '306',
  /**
   * The server sends this response to direct the client to get the
   * requested resource at another URI with same method that was used
   * in the prior request.
   *
   * This has the same semantics as the 302 Found HTTP response code,
   * with the exception that the user agent must not change the
   * HTTP method used: if a `POST` was used in the first request, a
   * `POST` must be used in the second request.
   */
  TEMPORARY_REDIRECT = '307',
  /**
   * This means that the resource is now permanently located at
   * another URI, specified by the Location: HTTP Response header.
   *
   * This has the same semantics as the 301 Moved Permanently
   * HTTP response code, with the exception that the user agent
   * must not change the HTTP method used: if a `POST` was used in the
   * first request, a `POST` must be used in the second request.
   */
  PERMANENT_REDIRECT = '308',
}


src/QuickRouter/handler.ts
/**
 * Copyright Dallen Baldwin 2022 - now
 * Distributed under the Boost Software License, Version 1.0.
 * (See accompanying file ./LICENSE or copy at https://www.boost.org/LICENSE_1_0.txt)
 */

import { RequestHandler, Request } from 'express';
import { defaultExtras } from '../util/defaultExtras';
import { ErrorCode } from '../constants/errors';
import { QuickRouterError } from '../QuickRouterError';
import { Success, SuccessCode } from '../constants/successes';

export interface HideDefaultExtras {
  /**
   * hide the request `params` object
   */
  params?: boolean;
  /**
   * hide the request `query` object
   */
  query?: boolean;
  /**
   * hide the request `body` object
   */
  body?: boolean;
}

export interface OnError {
  /**
   * a custom error code
   *
   * defaults to `ServerError.INTERNAL_SERVER_ERROR`
   */
  code?: ErrorCode;
  /**
   * a custom message to send back in the response
   */
  message: string;
  /**
   * any additional metadata to include in the response
   */
  extras?: Record<string, any>;
  /**
   * don't include default extras with the provided `extras` or at all
   * if no `extras` are provided
   */
  hideDefaultExtras?: boolean | HideDefaultExtras;
}

export interface OnErrorInput {
  /**
   * the request object
   */
  req: Request;
  /**
   * the error returned
   */
  err: unknown;
}

export interface HandlerOptions {
  /**
   * logic to execute as part of the response
   *
   * automatically wrapped in a `try/catch`
   */
  handler: (req: Request) => unknown | Promise<unknown>;
  /**
   * a custom code to send on success
   *
   * defaults to `200`
   */
  code?: SuccessCode;
  /**
   * custom logic to implement for errors
   *
   * if you want the error to fall through again, return `undefined`
   *
   * by default
   * - uses `500` for the code
   * - returns `message` from `ServerUtil.toError`
   * - returns `body`, `params`, `query` from `req` in `extras`
   */
  onError?: (input: OnErrorInput) => OnError | undefined;
}

/**
 * executes a request handler defined in the `HandlerOptions`
 *
 * if an error is thrown during the execution of `HandlerOptions.handler`,
 * it will be automatically parsed as a `QuickRouterError`
 *
 * if `HandlerOptions.onError` is defined any provided `OnError` properties
 * will modify the returned error
 *
 * @param options `HandlerOptions` to configure how responses are handled
 */
export const handler =
  (options: HandlerOptions): RequestHandler =>
  async (req, res) => {
    const { handler, code, onError } = options;
    let response: any;
    let statusCode: SuccessCode | ErrorCode = code ?? Success.OK;
    try {
      const data = await handler(req);
      response = { data };
    } catch (err) {
      let { message, traces, stack, code, extras } =
        QuickRouterError.fromUnknown(err);
      statusCode = code;
      extras = { ...defaultExtras(req), ...extras };
      if (onError)
        try {
          // catch errors returned in `onError`
          const custom = onError({ req, err });
          if (custom?.message) message = custom.message;
          if (custom?.code) statusCode = custom.code;
          if (custom?.extras) extras = { ...extras, ...custom.extras };
          if (custom?.hideDefaultExtras) {
            if (custom.hideDefaultExtras === true) extras = custom.extras;
            else {
              const { body, params, query } = custom.hideDefaultExtras;
              if (body && extras['body']) extras['body'] = undefined;
              if (params && extras['params']) extras['params'] = undefined;
              if (query && extras['query']) extras['query'] = undefined;
            }
          }
        } catch (err) {
          const parsed = QuickRouterError.fromUnknown(err);
          message = parsed.message;
          traces = parsed.traces;
          stack = parsed.stack;
        }
      response = { message, extras, traces, stack };
    } finally {
      res.status(+statusCode).send(response);
    }
  };
